spec/unit/typesafe/config/tokenizer_spec.rb

```diff
@@ -0,0 +1,696 @@
+  end
+
+  context "tokenize open and close curly braces" do
+    let(:test_string) { "{}" }
+    let(:expected_tokens) { [Tokens::OPEN_CURLY, Tokens::CLOSE_CURLY] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize open and close curly braces" do
+    let(:test_string) { "{}" }
+    let(:expected_tokens) { [Tokens::OPEN_CURLY, Tokens::CLOSE_CURLY] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize open square brackets" do
+    let(:test_string) { "[[" }
+    let(:expected_tokens) { [Tokens::OPEN_SQUARE, Tokens::OPEN_SQUARE] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize close square brackets" do
+    let(:test_string) { "]]" }
+    let(:expected_tokens) { [Tokens::CLOSE_SQUARE, Tokens::CLOSE_SQUARE] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize open and close square brackets" do
+    let(:test_string) { "[]" }
+    let(:expected_tokens) { [Tokens::OPEN_SQUARE, Tokens::CLOSE_SQUARE] }
+
+    include_examples "token_matching"
+  end
+
+  ####################
+  # comma, colon, equals, plus equals
+  ####################
+  context "tokenize comma" do
+    let(:test_string) { "," }
+    let(:expected_tokens) { [Tokens::COMMA] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize colon" do
+    let(:test_string) { ":" }
+    let(:expected_tokens) { [Tokens::COLON] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize equals" do
+    let(:test_string) { "=" }
+    let(:expected_tokens) { [Tokens::EQUALS] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize plus equals" do
+    let(:test_string) { "+=" }
+    let(:expected_tokens) { [Tokens::PLUS_EQUALS] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize comma, colon, plus equals, and equals together" do
+    let(:test_string) { "=:,+=" }
+    let(:expected_tokens) { [Tokens::EQUALS,
+                             Tokens::COLON,
+                             Tokens::COMMA,
+                             Tokens::PLUS_EQUALS] }
+
+    include_examples "token_matching"
+  end
+
+
+  ####################
+  # Substitutions
+  ####################
+  context "tokenize substitution" do
+    let(:test_string) { "${a.b}" }
+    let(:expected_tokens) { [TestUtils.token_substitution([TestUtils.token_unquoted("a.b")])] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize optional substitution" do
+    let(:test_string) { "${?x.y}" }
+    let(:expected_tokens) { [TestUtils.token_optional_substitution(TestUtils.token_unquoted("x.y"))] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize key substitution" do
+    let(:test_string) { "${?x.y}" }
+    let(:expected_tokens) { [TestUtils.token_key_substitution(TestUtils.token_unquoted("x.y"))] }
+
+    include_examples "token_matching"
+  end
+
+  ####################
+  # Unicode
+  ####################
+  context "tokenize unicode character" do
```

Comment:
Might be worth moving this below into the `Escaped Strings` section, but this is a nitpick, so feel free to ignore.